src/core/FdirSourceFileProvider.ts

```diff
@@ -0,0 +1,189 @@
+import { SourceFileProvider } from './SourceFileProvider';
+import { fdir } from 'fdir';
+import NormalizedPath from '../types/NormalizedPath';
+import * as ts from 'typescript';
+import { promisify } from 'util';
+import * as fs from 'fs';
+import * as path from 'path';
+const readFile = promisify(fs.readFile);
+const stat = promisify(fs.stat);
+import { createMatchPathAsync, MatchPathAsync } from 'tsconfig-paths';
+import { getScriptFileExtensions } from '../utils/getScriptFileExtensions';
+
+export class FDirSourceFileProvider implements SourceFileProvider {
+    parsedCommandLine: ts.ParsedCommandLine;
+    matchPath: MatchPathAsync;
+
+    constructor(configFileName: NormalizedPath, private rootDirs: string[]) {
+        // Load the full config file, relying on typescript to recursively walk the "extends" fields,
+        // while stubbing readDirectory calls to stop the full file walk of the include() patterns.
+        //
+        // We do this because we need to access the parsed compilerOptions, but do not care about
+        // the full file list.
+        this.parsedCommandLine = ts.getParsedCommandLineOfConfigFile(
+            configFileName,
+            {}, // optionsToExtend
+            {
+                getCurrentDirectory: process.cwd,
+                fileExists: fs.existsSync,
+                useCaseSensitiveFileNames: true,
+                readFile: path => fs.readFileSync(path, 'utf-8'),
+                readDirectory: () => {
+                    // this is supposed to be the recursive file walk.
+                    // since we don't care about _actually_ discovering files,
+                    // only about parsing the config's compilerOptions
+                    // (and tracking the "extends": fields across multiple files)
+                    // we short circuit this.
+                    return [];
+                },
+                onUnRecoverableConfigFileDiagnostic: diagnostic => {
+                    console.error(diagnostic);
+                    process.exit(1);
+                },
+            }
+        );
+
+        this.matchPath = createMatchPathAsync(
+            this.parsedCommandLine.options.baseUrl,
+            this.parsedCommandLine.options.paths
+        );
+    }
+
+    async getSourceFiles(searchRoots?: string[]): Promise<string[]> {
+        const allRootsDiscoveredFiles: string[][] = await Promise.all(
+            (searchRoots || this.rootDirs).map(
+                (rootDir: string) =>
+                    new fdir()
+                        .glob(
+                            this.parsedCommandLine.options.allowJs
+                                ? `**/*@(.js|.ts${

```

Comment:
Nit: for clarity, maybe you can precalculate the extensions to glob for and `Array.join` them into the glob expression here?  (Also has the nominal benefit of not recalculating it for every rootDir.)

Reply:
I think for consistency here I should update getScriptFileExtensions to do this calculation and call it here & below.

Reply:
I also end up needing a list of file extensions in #101, so I think deriving them from the same place is probably for the best.